hashing/src/lib.rs

```diff
@@ -1,174 +1,165 @@
+    serde(deny_unknown_fields)
 )]
-#[serde(deny_unknown_fields)]
-#[schemars(with = "String", description = "Hex-encoded hash digest.")]
 pub struct Digest(
-    #[serde(with = "HexForm::<[u8; Digest::LENGTH]>")]
-    #[schemars(skip, with = "String")]
+    #[cfg_attr(
+        feature = "std",
+        schemars(skip, with = "String"),
+        serde(with = "HexForm::<[u8; Digest::LENGTH]>")
+    )]
     [u8; Digest::LENGTH],
 );
 
 impl Digest {
-    /// Length of `Digest` in bytes.
     pub const LENGTH: usize = 32;
 
-    /// Returns a copy of the wrapped `u8` array.
-    pub fn to_array(self) -> [u8; Digest::LENGTH] {
+    /// Creates a 32-byte hash digest from a given a piece of data
+    pub fn hash<T: AsRef<[u8]>>(data: T) -> Digest {
+        let mut result = [0; Digest::LENGTH];
+
+        let mut hasher = VarBlake2b::new(Digest::LENGTH).expect("should create hasher");
+        hasher.update(data);
+        hasher.finalize_variable(|slice| {
+            result.copy_from_slice(slice);
+        });
+        Digest(result)
+    }
+
+    /// Hashes a pair of byte slices into a single [`Digest`]
+    pub fn hash_pair<T: AsRef<[u8]>, U: AsRef<[u8]>>(data1: T, data2: U) -> Digest {
+        let mut result = [0; Digest::LENGTH];
+        let mut hasher = VarBlake2b::new(Digest::LENGTH).unwrap();
+        hasher.update(data1);
+        hasher.update(data2);
+        hasher.finalize_variable(|slice| {
+            result.copy_from_slice(slice);
+        });
+        Digest(result)
+    }
+
+    /// Returns a reference to the underlying value.
+    pub fn inner(&self) -> &[u8; Digest::LENGTH] {
+        &self.0
+    }
+
+    /// Returns the underlying value.
+    pub fn value(&self) -> [u8; Digest::LENGTH] {
         self.0
     }
 
-    /// Returns a copy of the wrapped `u8` array as a `Vec`
+    /// Converts the underlying hash digest array to a `Vec`
+    #[cfg(feature = "std")]
     pub fn into_vec(self) -> Vec<u8> {
         self.0.to_vec()
     }
 
     /// Returns a `Digest` parsed from a hex-encoded `Digest`.
-    pub fn from_hex<T: AsRef<[u8]>>(hex_input: T) -> Result<Self, Error> {
+    #[cfg(feature = "std")]
+    pub fn from_hex<T: AsRef<[u8]>>(hex_input: T) -> Result<Self, hex::FromHexError> {
         let mut inner = [0; Digest::LENGTH];
         hex::decode_to_slice(hex_input, &mut inner)?;
         Ok(Digest(inner))
     }
-
-    /// Generates a random instance using a `TestRng`.
-    #[cfg(test)]
-    pub fn random(rng: &mut TestRng) -> Self {
-        Digest(rng.gen::<[u8; Digest::LENGTH]>())
-    }
 }
 
-impl AsRef<[u8]> for Digest {
-    fn as_ref(&self) -> &[u8] {
-        self.0.as_ref()
-    }
-}
-
-impl From<[u8; Digest::LENGTH]> for Digest {
-    fn from(inner: [u8; Digest::LENGTH]) -> Self {
-        Digest(inner)
+#[cfg(feature = "std")]
+impl std::fmt::LowerHex for Digest {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let hex_string = base16::encode_lower(self.inner());
```

Comment:
If we just iterate over `self.inner()` and write each byte we could avoid the string allocation.